server/portal/apps/projects/workspace_operations/shared_workspace_operations.py

```diff
@@ -0,0 +1,334 @@
+    client = service_account()
+    system_id = settings.PORTAL_PROJECTS_ROOT_SYSTEM_NAME
+    path = f"{workspace_id}"
+    client.files.mkdir(systemId=system_id, path=path)
+    return path
+
+
+def create_workspace_system(client, workspace_id: str, title: str) -> str:
+    system_id = f"{settings.PORTAL_PROJECTS_SYSTEM_PREFIX}.{workspace_id}"
+    system_args = {
+        "id": system_id,
+        "host": settings.PORTAL_PROJECTS_ROOT_HOST,
+        "port": int(settings.PORTAL_PROJECTS_SYSTEM_PORT),
+        "systemType": "LINUX",
+        "defaultAuthnMethod": "PKI_KEYS",
+        "canExec": False,
+        "rootDir": f"{settings.PORTAL_PROJECTS_ROOT_DIR}/{workspace_id}",
+        "effectiveUserId": settings.PORTAL_ADMIN_USERNAME,
+        "authnCredential": {
+            "privateKey": settings.PORTAL_PROJECTS_PRIVATE_KEY,
+            "publicKey": settings.PORTAL_PROJECTS_PUBLIC_KEY
+        },
+        "notes": {"title": title}
+    }
+    client.systems.createSystem(**system_args)
+    return system_id
+
+
+def increment_workspace_count(force=None) -> int:
+    client = service_account()
+    root = settings.PORTAL_PROJECTS_ROOT_SYSTEM_NAME
+    root_sys = client.systems.getSystem(systemId=root)
+    new_count = int(root_sys.notes.count) + 1
+
+    # Allow manual setting of the increment.
+    if force:
+        new_count = force
+
+    client.systems.patchSystem(systemId=root,
+                               notes={"count": new_count})
+    return new_count
+
+
+##########################################
+# HIGH-LEVEL OPERATIONS TIED TO API ROUTES
+##########################################
+
+
+def create_shared_workspace(client: Tapis, title: str, owner: str):
+    """
+    Create a workspace system owned by user whose client is passed.
+    """
+    service_client = service_account()
+    workspace_number = increment_workspace_count()
+    workspace_id = f"{settings.PORTAL_PROJECTS_ID_PREFIX}-{workspace_number}"
+
+    # Service client creates directory and gives owner write permissions
+    create_workspace_dir(workspace_id)
+    set_workspace_acls(service_client,
+                       settings.PORTAL_PROJECTS_ROOT_SYSTEM_NAME,
+                       workspace_id,
+                       owner,
+                       "add",
+                       "writer")
+
+    # User creates the system and adds their credential
+    system_id = create_workspace_system(client, workspace_id, title)
+    # priv_key, pub_key = createKeyPair()
+    # register_public_key(owner,
+    #                     pub_key,
+    #                     system_id)
+    # create_system_credentials(client, owner, pub_key, priv_key, system_id)
+    # create_system_credentials(service_client,
+    #                           settings.PORTAL_ADMIN_USERNAME,
+    #                          settings.PORTAL_PROJECTS_PUBLIC_KEY,
+    #                           settings.PORTAL_PROJECTS_PRIVATE_KEY,
+    #                           system_id)
```

Comment:
Can we remove the commented code in this file, or add a note as to why it's commented?

Reply:
I'll add a note. I don't want to lose this code because it _works_ in the case where users have their own permissions on files, and hopefully somewhere down the line we figure out a way to make that happen.